File: packages/gbnf-experiment/data/20260909T023701Z_e94dca40/ported_implementation/src/rules-builder/rules-builder-types.ts
export enum InternalRuleType {
  CHAR = 'char',
  CHAR_ALT = 'char_alt',
  ALT = 'alt',
  CHAR_NOT = 'char_not',
  CHAR_RNG_UPPER = 'char_rng_upper',
  REFERENCE = 'reference',
  END = 'end',
}

export class InternalRuleDefChar {
  type = InternalRuleType.CHAR as const;
  value: number[];

  constructor(value: number[] = []) {
    this.value = [...value];
  }
}

export class InternalRuleDefCharAlt {
  type = InternalRuleType.CHAR_ALT as const;
  value: number;

  constructor(value: number) {
    this.value = value;
  }
}

export class InternalRuleDefAlt {
  type = InternalRuleType.ALT as const;
}

export class InternalRuleDefCharNot {
  type = InternalRuleType.CHAR_NOT as const;
  value: number[];

  constructor(value: number[] = []) {
    this.value = [...value];
  }
}

export class InternalRuleDefCharRngUpper {
  type = InternalRuleType.CHAR_RNG_UPPER as const;
  value: number;

  constructor(value: number) {
    this.value = value;
  }
}

export class InternalRuleDefReference {
  type = InternalRuleType.REFERENCE as const;
  value: number;

  constructor(value: number) {
    this.value = value;
  }
}

export class InternalRuleDefEnd {
  type = InternalRuleType.END as const;
}

export type InternalRuleDef =
  | InternalRuleDefChar
  | InternalRuleDefCharAlt
  | InternalRuleDefAlt
  | InternalRuleDefCharNot
  | InternalRuleDefCharRngUpper
  | InternalRuleDefReference
  | InternalRuleDefEnd;

export type InternalRuleDefCharOrAltChar =
  | InternalRuleDefChar
  | InternalRuleDefCharAlt;

export const isRuleDefAlt = (
  rule?: InternalRuleDef | null,
): rule is InternalRuleDefAlt => rule instanceof InternalRuleDefAlt;

export const isRuleDefRef = (
  rule?: InternalRuleDef | null,
): rule is InternalRuleDefReference => rule instanceof InternalRuleDefReference;

export const isRuleDefEnd = (
  rule?: InternalRuleDef | null,
): rule is InternalRuleDefEnd => rule instanceof InternalRuleDefEnd;

export const isRuleDefChar = (
  rule?: InternalRuleDef | null,
): rule is InternalRuleDefChar => rule instanceof InternalRuleDefChar;

export const isRuleDefCharNot = (
  rule?: InternalRuleDef | null,
): rule is InternalRuleDefCharNot => rule instanceof InternalRuleDefCharNot;

export const isRuleDefCharAlt = (
  rule?: InternalRuleDef | null,
): rule is InternalRuleDefCharAlt => rule instanceof InternalRuleDefCharAlt;

export const isRuleDefCharRngUpper = (
  rule?: InternalRuleDef | null,
): rule is InternalRuleDefCharRngUpper =>
  rule instanceof InternalRuleDefCharRngUpper;

File: proposed-projects/round-trip-experiment/reverse/20260909T023701Z_e94dca40/20260910T060603Z_391335fe/ported_implementation/gbnf/rules_builder/rules_builder_types.py
from __future__ import annotations

from enum import Enum
from typing import Union


class InternalRuleType(str, Enum):
    CHAR = "char"
    CHAR_ALT = "char_alt"
    ALT = "alt"
    CHAR_NOT = "char_not"
    CHAR_RNG_UPPER = "char_rng_upper"
    REFERENCE = "reference"
    END = "end"

    def __str__(self) -> str:
        return str(self.value)


class InternalRuleDefChar:
    def __init__(self, value: list[int] | None = None) -> None:
        self.type = InternalRuleType.CHAR
        self.value = list(value) if value is not None else []


class InternalRuleDefCharAlt:
    def __init__(self, value: int) -> None:
        self.type = InternalRuleType.CHAR_ALT
        self.value = value


class InternalRuleDefAlt:
    def __init__(self) -> None:
        self.type = InternalRuleType.ALT


class InternalRuleDefCharNot:
    def __init__(self, value: list[int] | None = None) -> None:
        self.type = InternalRuleType.CHAR_NOT
        self.value = list(value) if value is not None else []


class InternalRuleDefCharRngUpper:
    def __init__(self, value: int) -> None:
        self.type = InternalRuleType.CHAR_RNG_UPPER
        self.value = value


class InternalRuleDefReference:
    def __init__(self, value: int) -> None:
        self.type = InternalRuleType.REFERENCE
        self.value = value


class InternalRuleDefEnd:
    def __init__(self) -> None:
        self.type = InternalRuleType.END


InternalRuleDef = Union[
    InternalRuleDefChar,
    InternalRuleDefCharAlt,
    InternalRuleDefAlt,
    InternalRuleDefCharNot,
    InternalRuleDefCharRngUpper,
    InternalRuleDefReference,
    InternalRuleDefEnd,
]

InternalRuleDefCharOrAltChar = Union[InternalRuleDefChar, InternalRuleDefCharAlt]


def is_rule_def_alt(rule: object | None = None) -> bool:
    return isinstance(rule, InternalRuleDefAlt)


def is_rule_def_ref(rule: object | None = None) -> bool:
    return isinstance(rule, InternalRuleDefReference)


def is_rule_def_end(rule: object | None = None) -> bool:
    return isinstance(rule, InternalRuleDefEnd)


def is_rule_def_char(rule: object | None = None) -> bool:
    return isinstance(rule, InternalRuleDefChar)


def is_rule_def_char_not(rule: object | None = None) -> bool:
    return isinstance(rule, InternalRuleDefCharNot)


def is_rule_def_char_alt(rule: object | None = None) -> bool:
    return isinstance(rule, InternalRuleDefCharAlt)


def is_rule_def_char_rng_upper(rule: object | None = None) -> bool:
    return isinstance(rule, InternalRuleDefCharRngUpper)
